File: src/main/scala/dojo/ttt/swing/Button.scala
package dojo.ttt.swing

import javax.swing.JButton
import java.awt.Dimension
import dojo.ttt.{PlayerToken, Empty, Token, Position}

class Button(val position: Position) extends JButton {

  private var token: Token = Empty

  setFont(getFont.deriveFont(100f))
  setText(label)
  setPreferredSize(new Dimension(120, 120))

  def isFree = token == Empty

  def reset() {
    token = Empty
    setText(label)
    setEnabled(true)
  }

  def setPlayer(player: PlayerToken) {
    token = player
    setText(label)
    setEnabled(false)
  }

  def label: String = if (isFree) null else token.toString

}

File: src/main/java/dojo/ttt/swing/Button.java
package dojo.ttt.swing;/*
* Copyright 2011 Yaana Technologies. All rights reserved.
* Yaana Technologies PROPRIETARY/CONFIDENTIAL.
*/

import javax.swing.*;

import dojo.ttt.Player;
import dojo.ttt.Position;

public class Button extends JButton {

    private final Position position;
    private Player player;

    public Button(Position position) {
        this.position = position;
    }

    public Position getPosition() {
        return position;
    }

    public boolean isFree() {
        return player == null;
    }

    public void setPlayer(Player player) {
        this.player = player;
        updateButtonText();
        setEnabled(false);
    }

    public void reset() {
        player = null;
        updateButtonText();
        setEnabled(true);
    }

    private void updateButtonText() {
        if (player == null) {
            setText(null);
        } else {
            setText(player.toString());
        }
    }
}
